bin/parallel.py

```diff
@@ -29,67 +34,82 @@ def __init__(self, f, num_threads=True):
             signal.signal(signal.SIGINT, self._interrupt_handler)
 
     def _worker(self):
-        try:
-            while not self.stopping:
-                task = self.q.get()
-                if task is None:
+        while True:
+            with self.mutex:
+                if len(self.tasks) == 0:
+                    self.todo.wait_for(lambda: len(self.tasks) > 0 or self.abort or self.finish)
+                
+                if self.abort or len(self.tasks) == 0:
```

Comment:
replace `len == 0` by `self.finish` for consistency with the `or`s above

Reply:
we don't want to directly leave on `self.finish`, we still want to handle all items in the queue first but i made that more obvious by changing the code :)